components/ai_chat/content/browser/associated_link_content_unittest.cc

```diff
@@ -0,0 +1,264 @@
+
+  base::test::TestFuture<PageContent> future;
+  testable_content->GetContent(future.GetCallback());
+
+  const PageContent& result = future.Take();
+  EXPECT_EQ(result.content, test_content());
+  EXPECT_FALSE(result.is_video);
+}
+
+TEST_F(AssociatedLinkContentTest, MultipleGetContentCallsWithEmptyCache) {
+  auto associated_content =
+      CreateAssociatedLinkContent(test_url(), test_title());
+  associated_content->set_page_content_result(test_content(), false);
+
+  base::test::TestFuture<PageContent> future1;
+  base::test::TestFuture<PageContent> future2;
+
+  associated_content->GetContent(future1.GetCallback());
+  associated_content->GetContent(future2.GetCallback());
+
+  EXPECT_FALSE(future1.IsReady());
+  EXPECT_FALSE(future2.IsReady());
+
+  associated_content->DocumentOnLoadCompletedInPrimaryMainFrame();
+
+  const auto& [content1, is_video1] = future1.Take();
+  EXPECT_EQ(content1, test_content());
+  EXPECT_FALSE(is_video1);
+
+  const auto& [content2, is_video2] = future2.Take();
+  EXPECT_EQ(content2, test_content());
+  EXPECT_FALSE(is_video2);
+}
+
+TEST_F(AssociatedLinkContentTest, DidFinishNavigationWithError) {
+  auto associated_content =
+      CreateAssociatedLinkContent(test_url(), test_title());
+
+  base::test::TestFuture<PageContent> future;
+  associated_content->GetContent(future.GetCallback());
+
+  content::MockNavigationHandle mock_handle(test_url(), main_rfh());
+  mock_handle.set_has_committed(true);
+  mock_handle.set_is_error_page(true);
+  mock_handle.set_net_error_code(net::ERR_CONNECTION_FAILED);
+
+  associated_content->DidFinishNavigation(&mock_handle);
+
+  const PageContent& result = future.Take();
+  EXPECT_TRUE(result.content.empty());
+  EXPECT_FALSE(result.is_video);
+}
+
+TEST_F(AssociatedLinkContentTest, DidFinishNavigationNonPrimaryFrame) {
+  auto associated_content =
+      CreateAssociatedLinkContent(test_url(), test_title());
+
+  base::test::TestFuture<PageContent> future;
+  associated_content->GetContent(future.GetCallback());
+
+  content::MockNavigationHandle mock_handle(test_url(), main_rfh());
+  mock_handle.set_has_committed(true);
+  mock_handle.set_is_error_page(true);
+  mock_handle.set_is_in_primary_main_frame(false);
+
+  associated_content->DidFinishNavigation(&mock_handle);
+
+  task_environment()->RunUntilIdle();
+  EXPECT_FALSE(future.IsReady());
+}
+
+TEST_F(AssociatedLinkContentTest, DidFinishNavigationNotCommitted) {
+  auto associated_content =
+      CreateAssociatedLinkContent(test_url(), test_title());
+
+  base::test::TestFuture<PageContent> future;
+  associated_content->GetContent(future.GetCallback());
+
+  content::MockNavigationHandle mock_handle(test_url(), main_rfh());
+  mock_handle.set_has_committed(false);
+  mock_handle.set_is_error_page(true);
+  mock_handle.set_is_in_primary_main_frame(true);
+
+  associated_content->DidFinishNavigation(&mock_handle);
+
+  task_environment()->RunUntilIdle();
```

Comment:
why is RunUntilIdle needed when we explicitly call DidFinishNavigation?

Reply:
Because the `OneShotEvent` is posted to the SequencedTaskRunner, so this doesn't complete synchronously. For the other tests I can wait for the future to complete, but in this case, I'm expecting it to not complete.

Reply:
Currently these two test cases leaves callback uncalled, shouldn't we simulate timeout by fast forwarding time for both to reflect our handling?

Reply:
We have a separate test for that:
https://github.com/brave/brave-core/pull/30963/files#diff-b8a1a1dd03633da97cceb64bc1b8857abf1e657bca417c2ff7794fc48ea19375R248

Here I just wanted to test that the callback is not called when we call `DidFinishNavigation` with the wrong params.

I'd still want the `RunUntilIdle(); EXPECT_FALSE(future.IsReady())`, I think to make sure the `DidFinishNavigation` firing the callback, but we could also test the Timeout here, if you think its a good idea?